[C++ for OpenCL] Added misc missing clarifications for address spaces (#263)
KhronosGroup/OpenCL-Docs: cxx4opencl/address_spaces.txt

==== Casts

C-style casts follow rules of `OpenCL C v2.0 s6.5.5`. Conversions to generic
address space can be made by any {cpp} cast operator (as an implicit
conversion); converting from generic to named address space can only be done

> using dedicated `addrspace_cast` operator. Operator `addrspace_cast` can only

Reviewer: ```suggestion
using the dedicated `addrspace_cast` operator. Operator `addrspace_cast` can only
```

convert between address spaces and no other conversions are allowed to occur.
Note that conversions between `__constant` and any other other address space
are disallowed.

[source,cpp]
----------
int * genptr;

> __private float * ptrfloat = reinterpret_cast<__private float*>(genptr); //illegal.

Reviewer: ```suggestion
__private float * ptrfloat = reinterpret_cast<__private float*>(genptr); // illegal.
```

> __private float * ptrfloat = addrspace_cast<__private float*>(genptr); //illegal.

Reviewer: ```suggestion
__private float * ptrfloat = addrspace_cast<__private float*>(genptr); // illegal.
```

__private int * ptr = addrspace_cast<__private int*>(genptr); // legal.
float * genptrfloat = reinterpret_cast<float*>(ptr); // legal.
__constant int * constptr = addrspace_cast<__constant int*>(genptr); // illegal.
----------

[[references]]
==== References

Reference types can be qualified with an address space.

[source,cpp]
----------
__private int & ref = ...; // references int in __private address space.
----------

> By default references will refer to generic address space objects, except

Reviewer: ```suggestion
By default references refer to generic address space objects, except
```

for dependent types that are not template specializations (see
<<addrspace-deduction, _Deduction_>>). Address space compatibility checks
are performed when references are bound to values. The logic follows the
rules from address space pointer conversion (`OpenCL v2.0 s6.5.5`).

[[addrspace-deduction]]

> ==== Deduction / Default address space

Reviewer: ```suggestion
==== Deduction & Default address space
```

This section details what happens if address spaces for types are not

> provided in the source code explicitly. Most of the logic for the address space

Reviewer: ```suggestion
provided in the source code explicitly. Most of the logic for address space
```

deduction (i.e. default address space) follows rules from `OpenCL 2.0 s6.5`.

> References inherit rules from the pointers and therefore refer to generic

Reviewer: ```suggestion
References inherit rules from pointers and therefore refer to generic
```

==== Casts

> C-style casts follow rules from `OpenCL C v2.0 s6.5.5`. All cast operators

Reviewer 2: "rules from" -> "the rules of"

> permit conversion to generic address space implicitly. However, converting

Reviewer 2: , however,

Reviewer 2: This reads slightly strangely. Could we say "conversions to generic address space can be made by any cast operator; converting from generic to named ..."

==== Casts

C-style casts follow rules of `OpenCL C v2.0 s6.5.5`. Conversions to generic
address space can be made by any {cpp} cast operator (as an implicit
conversion); converting from generic to named address space can only be done
using dedicated `addrspace_cast` operator. Operator `addrspace_cast` can only
convert between address spaces and no other conversions are allowed to occur.
Note that conversions between `__constant` and any other other address space
are disallowed.

> [source,c]

Reviewer 3: I've never needed to try before so I don't know if it works.  But is `[source,cpp]` or similar supported, and does it result in any improvements?

Author: Works nicely! Thanks!

Author: Changed in https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/adeb8aae4ab18510f35a84cb6db53c4bbd716d32

----------
int * genptr;
__private float * ptrfloat = reinterpret_cast<__private float*>(genptr); //illegal.

> __private int * ptr = addrspace_cast<__private int*>(genptr);

Reviewer 3: ```suggestion
__private int * ptr = addrspace_cast<__private int*>(genptr); // legal
```

Would make this clearer to me, it's also what OpenCL C spec does for these examples.

> __private int * ptr = addrspace_cast<__private int*>(genptr);
> __private float * ptrfloat = addrspace_cast<__private float*>(genptr); //illegal.

Reviewer 4: Could you swap those two lines? It took me a while to see that the variable had a different name and was used below.

```suggestion
__private float * ptrfloat = addrspace_cast<__private float*>(genptr); //illegal.
__private int * ptr = addrspace_cast<__private int*>(genptr);
```

Author: Changed in https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/adeb8aae4ab18510f35a84cb6db53c4bbd716d32

are performed when references are bound to values. The logic follows the
rules from address space pointer conversion (`OpenCL v2.0 s6.5.5`).

[[addrspace-deduction]]
==== Deduction / Default address space

This section details what happens if address spaces for types are not

> provided in the source code explicitly. Most of logic for the address space

Reviewer 2: Most of *the* logic

> deduction (i.e. default address space) follow rules from `OpenCL 2.0 s6.5`.

Reviewer 2: follows rules

    except for template specializations.
  * non-pointer/non-reference class members except for static data members
    that are deduced to `__global` address space.
  * non-pointer/non-reference alias declarations.
  * decltype expressions.

[source,c]
----------
template <typename T>
void foo() {
  T m; // address space of 'm' will be known at template instantiation time.
  T * ptr; // 'ptr' points to generic address space object.
  T & ref = ...; // 'ref' references an object in generic address space.
};

template <int N>
struct S {
  int i; // 'i' has no address space.
  static int ii; // 'ii' is in global address space.
  int * ptr; // 'ptr' points to generic address space int.
  int & ref = ...; // 'ref' references int in generic address space.
};

template <int N>
void bar()
{
  S<N> s; // 's' is in __private address space.
}
----------

[source,c]
----------
struct c1 {};
using alias_c1 = c1; // 'alias_c1' is 'c1'.
using alias_c1_ptr = c1 *; // 'alias_c1_ptr' is a generic address space pointer to 'c1'.
----------

[source,c]
----------
__kernel void foo()
{
  __local int i;
  decltype(i)* ii; type of 'ii' is 'local int *__private'.
}
----------

> For the placeholder type specifiers `auto` an address space of the outer type is

Reviewer 2: specifier auto

deduced as if it would be any other regular type. However if auto is used in

> reference or a pointer type, the address space of a pointee is taken from the type

Reviewer 2: a reference or pointer type

> of initialization expression. The logic simply follows `const`/`volatile` qualifiers.

Reviewer 2: of the initialization

access coalescing for segments with memory bank. This not only applies
to regular member functions but to constructors and destructors too.

Address spaces are not deduced for:

  * non-pointer/non-reference template parameters or any dependent types
    except for template specializations.
  * non-pointer/non-reference class members except for static data members
    that are deduced to `__global` address space.
  * non-pointer/non-reference alias declarations.
  * decltype expressions.

[source,c]
----------
template <typename T>
void foo() {
  T m; // address space of 'm' will be known at template instantiation time.
  T * ptr; // 'ptr' points to generic address space object.
  T & ref = ...; // 'ref' references an object in generic address space.
};

> template <int N>

Reviewer 3: Is the fact that this struct is templated important to the deduction rules?  N is never relevant.

Author: I think the aim of this example is to show that the address space deduction is performed normally if the type of variable doesn't depend on the type of a template parameter. In this case it is using a value of template parameter and not a type.

Reviewer 3: Okay, thanks.

struct S {
  int i; // 'i' has no address space.
  static int ii; // 'ii' is in global address space.
  int * ptr; // 'ptr' points to generic address space int.
  int & ref = ...; // 'ref' references int in generic address space.
};

template <int N>
void bar()
{
  S<N> s; // 's' is in __private address space.
}
----------

[source,c]
----------
struct c1 {};
using alias_c1 = c1; // 'alias_c1' is 'c1'.
using alias_c1_ptr = c1 *; // 'alias_c1_ptr' is a generic address space pointer to 'c1'.
----------

[source,c]
----------
__kernel void foo()
{
  __local int i;

>   decltype(i)* ii; type of 'ii' is 'local int *__private'.

Reviewer 4: ```suggestion
  decltype(i)* ii; // type of 'ii' is '__local int *__private'.
```

}
----------

For the placeholder type specifier `auto` an address space of the outer type is
deduced as if it would be any other regular type. However if auto is used in a
reference or pointer type, the address space of a pointee is taken from the type

> of the initialization expression. The logic simply follows `const`/`volatile`

Reviewer 4: ```suggestion
of the initialization expression. The logic follows `const`/`volatile`
```

qualifiers.

[source,c]
----------
__kernel void foo()
{
  __local int i;
  constexpr int c = 1;
  
  __constant auto cai = c; // type of 'cai' is '__constant int' (no deduction).
  
  auto aii = cai; // type of 'aii' is '__private int' (regular deduction).

>   auto *ptr = &i; // type of 'ptr' is '__local int * __private'
>                 // (addr space of a pointer is deduced regularly,
>                 // addr space of its pointee is taken from 'i').

Reviewer 4: ```suggestion
  auto *ptr = &i; // type of 'ptr' is '__local int * __private'
                  // (addr space of a pointer is deduced regularly,
                  // addr space of its pointee is taken from 'i').
```

  * non-pointer/non-reference alias declarations.
  * decltype expressions.

[source,cpp]
----------
template <typename T>
void foo() {
  T m; // address space of 'm' will be known at template instantiation time.
  T * ptr; // 'ptr' points to generic address space object.
  T & ref = ...; // 'ref' references an object in generic address space.
};

template <int N>
struct S {
  int i; // 'i' has no address space.
  static int ii; // 'ii' is in global address space.

>   int * ptr; // 'ptr' points to generic address space int.

Reviewer: ```suggestion
  int * ptr; // 'ptr' points to int in generic address space.
```

  int & ref = ...; // 'ref' references int in generic address space.
};

template <int N>
void bar()
{
  S<N> s; // 's' is in __private address space.
}
----------

[source,cpp]
----------
struct c1 {};
using alias_c1 = c1; // 'alias_c1' is 'c1'.
using alias_c1_ptr = c1 *; // 'alias_c1_ptr' is a generic address space pointer to 'c1'.
----------

[source,cpp]
----------
__kernel void foo()
{
  __local int i;
  decltype(i)* ii; // type of 'ii' is '__local int *__private'.
}
----------

For the placeholder type specifier `auto` an address space of the outer type is

> deduced as if it would be any other regular type. However if auto is used in a

Reviewer: ```suggestion
deduced as if it would be any other regular type. However if `auto` is used in a
```

reference or pointer type, the address space of a pointee is taken from the type

> of the initialization expression. The logic follows `const`/`volatile`

Reviewer: ```suggestion
of the initialization expression. The logic follows `const` and `volatile`
```

Author: https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/5aa92f05b8470cc9aaee07a1d29b02d5d84d8bc9

> qualifiers.

Reviewer: ```suggestion
qualifier logic.
```

Author: There is already "logic" at the beginning of this sentence. I think I will change to rules.

Reviewer: That's fine with me too.

Author: https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/079a393b2ba07a6d68f1292a15b93aa514cd2276

}
----------

==== Lambda function

> The address space qualifier can be added optionally for lambda

Reviewer 2: optionally added

}
----------

==== Lambda function

The address space qualifier can be optionally added for lambda

> expressions after the attributes. Similarly to method qualifiers

Reviewer: ```suggestion
expressions after the attributes. Similar to method qualifiers,
```

}
----------

==== Lambda function

> The address space qualifier can be optionally added for lambda

Reviewer 4: ```suggestion
The address space qualifier can be optionally added to lambda
```

I think...

expressions after the attributes. Similarly to method qualifiers
they will alter the default address space of lambda call operator
that has generic address space by default. 

[source,c]
----------
__kernel void test_qual() {
  auto priv1 = []() __private {};
  priv1();

>   auto priv3 = []() __global {};
>   priv3(); // error: lambda object and its expression have mismatching address space.

Reviewer 4: ```suggestion
  auto priv2 = []() __global {};
  priv2(); // error: lambda object and its expression have mismatching address space.
```

to keep numbering of variable consistent in the example.

}
----------

==== Construction, initialization and destruction

Construction, initialization and destruction of objects in `__private`

> and `__global` address space follows general principals of C++. For the

Reviewer 2: follows the general

program scope objects the implementation (i.e. compiler) defines an ABI
format for initialization and destruction of global objects before/after
all kernels are enqueued.

> Objects in `__local` address space can not have initializers in

Reviewer 2: in the `local`

declarations and therefore a constructor can not be called. All objects

> created in local address space have undefined state at the beginning.

Reviewer 2: created in the local

Developers are free to defined a special member function that can
initialize local address space objects after their declaration. Any
default values provided for the initialization of members in a class
declaration are ignored when creating the local address space objects.
Classes that are instantiated with local address space objects can not
have references as their members. The same restriction applies recursively
to the members of such classes. Destructors of local address space objects
are not invoked automatically. They can be called manually if required.

User defined constructors are not allowed to construct objects in `__constant`

> address space. Such objects can be however initialized using

Reviewer 2: drop however here

}
----------

==== Construction, initialization and destruction

Construction, initialization and destruction of objects in `__private`

> and `__global` address space follows the general principals of {cpp}. For

Reviewer 4: ```suggestion
and `__global` address space follows the general principles of {cpp}. For
```

> and `__global` address space follows the general principles of {cpp}. For

Reviewer: ```suggestion
and `__global` address space follow the general principles of {cpp}. For
```

> the program scope objects the implementation (i.e. compiler) defines an ABI

Reviewer: ```suggestion
program scope objects, the implementation (i.e. compiler) defines an ABI
```

format for initialization and destruction of global objects before/after
all kernels are enqueued.

Objects in `__local` address space can not have initializers in
declarations and therefore a constructor can not be called. All objects

> created in the local address space have undefined state at the beginning.

Reviewer 3: Could say something more  precise that "beginning".   I suspect that it means "at the start of kernel execution".

Author: Yes, or at the point of declaration... not sure which one is more precise... or perhaps there is no difference...

Author: Changed in https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/adeb8aae4ab18510f35a84cb6db53c4bbd716d32

> their declaration. Developers are free to defined a special member function

Reviewer: ```suggestion
their declaration. Developers are free to define a special member function
```

initialize local address space objects after their declaration. Any
default values provided for the initialization of members in a class
declaration are ignored when creating the local address space objects.
Classes that are instantiated with local address space objects can not
have references as their members. The same restriction applies recursively

> to the members of such classes. Destructors of local address space objects

Reviewer 4: I suppose this applies to base classes as well?

```suggestion
to the members and base classes. Destructors of local address space objects
```

Author: Absolutely! Thanks!
